geode-core/src/main/java/org/apache/geode/cache/wan/GatewaySender.java

```diff
@@ -174,7 +174,32 @@
    */
   int GET_TRANSACTION_EVENTS_FROM_QUEUE_RETRIES =
       Integer.getInteger(GeodeGlossary.GEMFIRE_PREFIX + "get-transaction-events-from-queue-retries",
-          10);
+          2);
```

Comment:
Prior to this change, retries were done immediately if it was detected that not all events for a transaction were in the batch. In this change, I have added a configurable sleep before retrying with a default value of 1ms because I saw that retries were sometimes done so fast that it was not enough to get the expected events in the queue and it was more reasonable to have less retries and have some waiting before retrying.

Reply:
But 2 is still too less. How about 3~5?

Reply:
> But 2 is still too less. How about 3~5?

Given that prior to this change we had 10 retries and no timeout, and these retries were done really fast (in less than 1 ms) I did not want to add a much higher waiting time because it does not seem necessary if the events will eventually reach the queue so I thought it would be better to have some small waiting time (2ms) and have a smaller number of retries. These numbers seem reasonable and have not caused flakyneess in the test cases.

If we put a higher waiting time and more retries as you are suggesting and, for some reason, the configuration of the gateway senders or the traffic received contains transactions but they cannot be completed (even with infinite time), then you are going to be delaying replication for a long time: each batch for waiting time*number of retries which could be a lot.

That's why I prefer to leave these values low and, if in some case it is necessary, they can be changed with the system properties.